File: exercises/day11-AWorldWithoutMock/java/src/test/java/service/ToyProductionServiceTest.java
package service;

import domain.Toy;
import domain.ToyRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.InOrder;
import org.mockito.junit.jupiter.MockitoExtension;

import static domain.Toy.State.IN_PRODUCTION;
import static domain.Toy.State.UNASSIGNED;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class ToyProductionServiceTest {

    private static final String TOY_NAME = "Train";

    @Mock
    ToyRepository repository;

    @Mock
    NotificationService notificationService;

    @InjectMocks
    ToyProductionService service;

    @Test
    void assignToyToElfShouldSaveToyInProductionAndNotify() {
        // given
        Toy toy = new Toy(TOY_NAME, UNASSIGNED);
        when(repository.findByName(TOY_NAME)).thenReturn(toy);

        // when
        service.assignToyToElf(TOY_NAME);

        // then
        ArgumentCaptor<Toy> toyCaptor = ArgumentCaptor.forClass(Toy.class);
        verify(repository).save(toyCaptor.capture());
        Toy savedToy = toyCaptor.getValue();
        assertThat(savedToy.getState()).isEqualTo(IN_PRODUCTION);

        verify(notificationService).notifyToyAssigned(savedToy);
    }

    @Test
    void assignToyToElfShouldNotSaveOrNotifyWhenToyNotFound() {
        // given
        when(repository.findByName(TOY_NAME)).thenReturn(null);

        // when
        service.assignToyToElf(TOY_NAME);

        // then
        verify(repository).findByName(TOY_NAME);
        verify(repository, never()).save(any());
        verifyNoInteractions(notificationService);
    }

    @Test
    void assignToyToElfShouldNotSaveOrNotifyWhenToyAlreadyInProduction() {
        // given
        Toy toy = new Toy(TOY_NAME, IN_PRODUCTION);
        when(repository.findByName(TOY_NAME)).thenReturn(toy);

        // when
        service.assignToyToElf(TOY_NAME);

        // then
        verify(repository).findByName(TOY_NAME);
        verify(repository, never()).save(any());
        verifyNoInteractions(notificationService);
    }

    @Test
    void assignToyToElfShouldSaveBeforeNotifying() {
        // given
        Toy toy = new Toy(TOY_NAME, UNASSIGNED);
        when(repository.findByName(TOY_NAME)).thenReturn(toy);

        // when
        service.assignToyToElf(TOY_NAME);

        // then
        InOrder inOrder = inOrder(repository, notificationService);
        inOrder.verify(repository).findByName(TOY_NAME);
        inOrder.verify(repository).save(any(Toy.class));
        inOrder.verify(notificationService).notifyToyAssigned(any(Toy.class));
        inOrder.verifyNoMoreInteractions();
    }
}

File: exercises/day11-AWorldWithoutMock/kotlin/src/test/kotlin/service/ToyProductionServiceTest.kt
package service

import domain.Toy
import domain.ToyRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*

@ExtendWith(MockitoExtension::class)
class ToyProductionServiceTest {

    companion object {
        private const val TOY_NAME = "Train"
    }

    @Mock
    lateinit var repository: ToyRepository

    @Mock
    lateinit var notificationService: NotificationService

    @InjectMocks
    lateinit var service: ToyProductionService

    @Test
    fun `assignToyToElf should save toy in production and notify`() {
        val toy = Toy(TOY_NAME, Toy.State.UNASSIGNED)
        whenever(repository.findByName(TOY_NAME)).thenReturn(toy)

        service.assignToyToElf(TOY_NAME)

        val toyCaptor = argumentCaptor<Toy>()
        verify(repository).save(toyCaptor.capture())
        val savedToy = toyCaptor.firstValue
        assertThat(savedToy.state).isEqualTo(Toy.State.IN_PRODUCTION)

        verify(notificationService).notifyToyAssigned(savedToy)
    }

    @Test
    fun `assignToyToElf should not save or notify when toy not found`() {
        whenever(repository.findByName(TOY_NAME)).thenReturn(null)

        service.assignToyToElf(TOY_NAME)

        verify(repository).findByName(TOY_NAME)
        verify(repository, never()).save(any())
        verifyNoInteractions(notificationService)
    }

    @Test
    fun `assignToyToElf should not save or notify when toy already in production`() {
        val toy = Toy(TOY_NAME, Toy.State.IN_PRODUCTION)
        whenever(repository.findByName(TOY_NAME)).thenReturn(toy)

        service.assignToyToElf(TOY_NAME)

        verify(repository).findByName(TOY_NAME)
        verify(repository, never()).save(any())
        verifyNoInteractions(notificationService)
    }

    @Test
    fun `assignToyToElf should save before notifying`() {
        val toy = Toy(TOY_NAME, Toy.State.UNASSIGNED)
        whenever(repository.findByName(TOY_NAME)).thenReturn(toy)

        service.assignToyToElf(TOY_NAME)

        inOrder(repository, notificationService) {
            verify(repository).findByName(TOY_NAME)
            verify(repository).save(any())
            verify(notificationService).notifyToyAssigned(any())
            verifyNoMoreInteractions()
        }
    }
}